### PR TITLE
feat!: Implement a DBScan clustering algorithm directly in ACTS

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
With this branch we removes all dependencies on MLPack for the clustering used in the ML seed filtering algorithm and replaces them with a custom implementation of the DBScan density-based clustering algorithm.

This new algorithm is far more powerful than the previous MLPack implementation and supports the full C++. With it, we observe a speed-up of the seed filtering algorithm by a factor of 2 and remove external dependency. All tracking performances should stay unchanged even if some minor difference might appear at the border of 2 nearby clusters (since the entries are not processed in the same order compared to the old mlpack implementation)

### CODE REVIEW

File: Core/include/Acts/Utilities/DBScan.hpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  /// @brief Extend the cluster.
+  ///
+  /// This function will extend the cluster by finding all the neighbours of the
+  /// current point and assign them to the current cluster.
+  /// The KDTree is used to find the neighbours and an extra cut is used to only
+  /// keep the neighbours that are within the epsilon radius.
+  ///
+  /// @param tree The KDTree containing all the points.
+  /// @param inputPoints The vector containing the input points.
+  /// @param pointToProcess The vector containing the ids of the points that need to be
```
Comment (AJPfleger) [inline Core/include/Acts/Utilities/DBScan.hpp:142]:
```suggestion
  /// @param pointsToProcess The vector containing the ids of the points that need to be
```
Maybe change the variable `pointToProcess` to `pointsToProcess`, to show that we are dealing with a set of points.

File: Examples/Algorithms/TrackFindingML/include/ActsExamples/TrackFindingML/SeedFilterDBScanClustering.hpp [hunk 2]
```diff
@@ -1,61 +1,47 @@
...
 namespace Acts {
 
 /// Clusters seed based on their direction, their Z impact parameter and their
 /// momentum using DBScan
 ///
 /// @param input : Input parameters for the clustering (phi, eta, z, Pt)
-/// @param epsilon : Maximum distance between 2 tracks to be clustered
-/// @param minPoints : Minimum number of tracks to create a cluster
+/// @param epsilon : Maximum distance between 2 seed to be clustered
+/// @param minPoints : Minimum number of seed to create a cluster
```
Comment (AJPfleger) [inline Examples/Algorithms/TrackFindingML/include/ActsExamples/TrackFindingML/SeedFilterDBScanClustering.hpp:24]:
```suggestion
/// @param epsilon Maximum distance between 2 seeds to be clustered
/// @param minPoints Minimum number of seeds to create a cluster
```
could you also remove the `: ` in the line above? :)

File: Tests/UnitTests/Core/Utilities/DBScanTests.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <iterator>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace {
+std::vector<std::array<double, 2>> test_vector{
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Utilities/DBScanTests.cpp:22]: Do the values have any special meaning or completely random?
Maybe you could add tiny description like
```
A test vector with random values. They form 4 clusters around ... and should be separable with an epsilon of ...
```

**Review by AJPfleger — CHANGES_REQUESTED** [review]
I made some nit picky comments regarding the naming convention and some optional `const` which could make the code more readable. I didn't mark all occurrences of the variables, where I suggested the changes.
The review got a bit cluttered already in `DBScan.hpp`. Maybe you could look over the suggestions and after that I have second look.

File: Core/include/Acts/Utilities/DBScan.hpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  /// processed.
+  /// @param clusteredPoints Vector containing the cluster ID of each point.
+  /// @param clusterID The ID of the current cluster.
+  ///
+  void expendCluster(const tree_t& tree, const vector_p& inputPoints,
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Core/include/Acts/Utilities/DBScan.hpp:147]: The original DBScan paper uses expand (with an a for this function). So I will go with that if it fine by you ?

File: Core/include/Acts/Utilities/DBScan.hpp [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                     const std::vector<std::size_t>& pointToProcess,
+                     std::vector<int>& clusteredPoints, int clusterID) {
+    // Loop over all the points that need to be process.
+    for (auto& id : pointToProcess) {
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Core/include/Acts/Utilities/DBScan.hpp:151]: I agree with the const but I will need to test that the speed doesn't change by removing the ref (I have seen some large changes by doing change like that in other part of the code)

File: Core/include/Acts/Utilities/DBScan.hpp (hunk 1 above)
Comment (Corentin-Allaire) [reply Core/include/Acts/Utilities/DBScan.hpp:142]: good idea

File: Tests/UnitTests/Core/Utilities/DBScanTests.cpp (hunk 3 above)
Comment (Corentin-Allaire) [reply Tests/UnitTests/Core/Utilities/DBScanTests.cpp:22]: I took them from a ML lecture I gave last year. They were generated in python from the sklearn `make_blobs` function to correspond to 4 cluster. I will add a comment

**Comment by Corentin-Allaire** [comment]
So I should have fixed most of this. But I will be doing some test to ensure that the speed hasn't changed too much

**Comment by Corentin-Allaire** [comment]
On my machine we are staying at a consistent 100ms per pu200 ttbar event so this is good !

**Review by AJPfleger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [7f6f1314c4fe544b086e103eecff5c703683939a]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259935)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259934)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259933)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259932)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259931)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259930)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259929)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259928)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259927)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259926)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259925)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259924)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259923)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259922)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259921)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259920)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259919)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259918)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259917)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259916)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259915)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259914)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/35259913)||

**Pull request merged**